### ISSUE TITLE
Inconsistent quotation marks sanitization behavior when JSONata functions are involved

### ISSUE THREAD

**Issue by raymond-chen-ampl** [IBM/JSONata4Java#237]
## Description
when functions like `$each(...)` and `$map(...)` are involved, strings with surrounded double quotation marks will get sanitized twice accidentally 
Here's a small example to reproduce:
```
public static void main(String[] args) throws Exception {
        ObjectMapper mapper = new ObjectMapper();

        var payload = mapper.readTree("""
                { "key1": "\\"v1\\"" }
                """);
        var result1 = Expressions.parse("$each($, function($v, $k) {({ $k: $v})}) ~> $merge()").evaluate(payload);
        var result2 = Expressions.parse("{ \"key1\": key1 }").evaluate(payload);

        System.out.println(mapper.writeValueAsString(result1));
        System.out.println(mapper.writeValueAsString(result2));
    }
```
This will output
```
{"key1":"v1"}
{"key1":"\"v1\""}
```

While in https://try.jsonata.org/, both will output `{"key1":"\"v1\""}`

## Root Cause
Functions like `$each()` and `$map()` internally use `FunctionUtils. fillExprVarContext(...)`, which makes use of Jackson's `asText()` to build the context. 
https://github.com/IBM/JSONata4Java/blob/8e12caa93ae16aa2ed805abe074e243de3c1c046/src/main/java/com/api/jsonata4java/expressions/utils/FunctionUtils.java#L251

This is fine for all the other data types until it goes to String: Jackson handles quotation escape implicitly in `asText()` while not in `toString()`. 
https://stackoverflow.com/questions/48655074/what-is-the-difference-between-astext-and-tostring-in-jsonnode

Therefore, using `asText` will lead to a double sanitization in `visitString` 

I believe we should use `toString()` here to keep the raw value all the way until `visitString`

**Comment by wnm3** [comment]
Thank you for your thorough explanation and solution. I've checked in an update to FunctionUtils.java with the change from asText to toString. Will hold off making a new release pending some other issues. However, you can generate jar files by issuing the command:  
mvn clean install -Dgpg.skip  
in the JSONata4Java directory. The jar files will be in the target subdirectory.

**Comment by bluemel** [comment]
I also like Raymond's proposal. I'm only a little bit unsure about the solution that's currently on master:

1) One question for me is if we should continue to keep case STRING and default together or if we better handle them separately from now on. At least if the test coverage is sufficient we could assume that the current solution on master is OK.

2) Additionally I found out that we probably could eliminate method:
   public static ExprValuesContext fillExprVarContext(ExprContext ctx, String key, JsonNode value)
   It seems to be completely covered by method:
   public static ExprValuesContext fillExprVarContext(int fctVarCount, ExprContext ctx, JsonNode... elements)

Anyway I'm preparing a pull request with another fix on the filter function which I intend to issue a pull request soon,

**Comment by wnm3** [comment]
I changed the code per Raymond's request and all tests passed. I added a test for his use case.

I'd be surprised we could remove the method as it takes different parameters. I suppose if you can marshall the parameters from how it is currently called into the other signature then we could do without it. But ExprContext is a tree, and I don't recall why we have the key parameter in one and not the other...

I'll wait on making a new release pending the outcome of your work on the new pull request and will not update main in the meantime.

**Comment by bluemel** [comment]
OK Nat, think my work has a releasable state so far.
See my pull request:
https://github.com/IBM/JSONata4Java/pull/238

**Comment by wnm3** [comment]
Okay, I'll merge this pull request #238 and produce a new version. You can approach additional work in another pull request. Thanks.

**Comment by wnm3** [comment]
@raymond-chen-ampl if you agree this issue can be closed, please close it. I've released version 2.2.5 today so will be in Maven Central in a couple of hours. Check here for version 2.2.5: https://mvnrepository.com/artifact/com.ibm.jsonata4java/JSONata4Java

**Comment by raymond-chen-ampl** [comment]
@wnm3 @bluemel I pulled version 2.2.5 and the issue was resolved. Thank you both for working on this !
